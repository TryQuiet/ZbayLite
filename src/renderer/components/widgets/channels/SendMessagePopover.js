import React from 'react'
import PropTypes from 'prop-types'
import * as R from 'ramda'
import Popover from '@material-ui/core/Popover'
import Jdenticon from 'react-jdenticon'
import { withStyles } from '@material-ui/core/styles'

import QuickActionLayout from '../../ui/QuickActionLayout'

const styles = theme => ({})

export const SendMessagePopover = ({
  classes,
  username,
  address,
  anchorEl,
  handleClose,
  isUnregistered,
  createNewContact,
  history,
  users
}) => {
  const open = Boolean(anchorEl)
  const id = open ? 'simple-popover' : undefined
  const registeredUsername = Array.from(Object.values(users))
    .filter(obj => obj.address === address)[0]
  return (
    <Popover
      className={classes.popover}
      id={id}
      open={open}
      anchorEl={anchorEl}
      onClose={handleClose}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'center'
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'center'
      }}
    >
      <QuickActionLayout
        main={username}
        buttonName='Send message'
        handleClose={handleClose}
        warrning={
          isUnregistered ? 'Unregistered users cannot receive messages.' : null
        }
        onClick={() => {
          createNewContact({
            contact: {
              address,
              nickname: username,
              publicKey: registeredUsername ? registeredUsername.publicKey : null
            },
            history
          })
        }
        }
      >
        <Jdenticon size='100' value={username} />
      </QuickActionLayout>
    </Popover>
  )
}

SendMessagePopover.propTypes = {
  classes: PropTypes.object.isRequired,
  username: PropTypes.string.isRequired,
  handleClose: PropTypes.func.isRequired,
  address: PropTypes.string.isRequired,
  anchorEl: PropTypes.bool,
  isUnregistered: PropTypes.bool
}

export default R.compose(
  React.memo,
  withStyles(styles)
)(SendMessagePopover)
