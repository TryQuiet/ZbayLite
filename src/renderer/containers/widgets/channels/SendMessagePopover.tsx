import React, { useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import SendMessagePopoverComponent from '../../../components/widgets/channels/SendMessagePopover'
// import identitySelectors from '../../../store/selectors/identity'
// import userSelectors from '../../../store/selectors/users'
import directMessages from '../../../store/handlers/contacts'
import { IUser } from '../../../store/handlers/directMessages'
import directMessagesSelectors from '../../../store/selectors/directMessages'

interface useSendMessagePopoverDataReturnType {
  users: any[]
  waggleUsers: {
    [key: string]: IUser
  }
}

export const useSendMessagePopoverData = (): useSendMessagePopoverDataReturnType => {
  const data = {
    // identityId: identitySelectors.id(state),
    // identityId: 'id',
    // users: userSelectors.users(state),
    users: [],
    waggleUsers: useSelector(directMessagesSelectors.users)
  }
  return data
}

export const useSendMessagePopoverActions = () => {
  const dispatch = useDispatch()

  const createNewContact = useCallback((contact) => {
    dispatch(directMessages.epics.createVaultContact(contact))
  }, [dispatch])

  return { createNewContact }
}

export const SendMessagePopover = ({ username, anchorEl, handleClose }) => {
  const { users, waggleUsers } = useSendMessagePopoverData()
  const { createNewContact } = useSendMessagePopoverActions()

  return (
    <SendMessagePopoverComponent
      username={username}
      anchorEl={anchorEl}
      users={users}
      waggleUsers={waggleUsers}
      createNewContact={createNewContact}
      handleClose={handleClose}
      isUnregistered={false}
    />
  )
}

export default SendMessagePopover
