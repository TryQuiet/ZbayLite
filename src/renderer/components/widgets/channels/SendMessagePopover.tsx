import React from 'react'
import Popover from '@material-ui/core/Popover'
import Jdenticon from 'react-jdenticon'
import { ISendMessagePopoverProps } from './SendMessagePopover.d'
import QuickActionLayout from '../../ui/QuickActionLayout/QuickActionLayout'

export const SendMessagePopover: React.FC<ISendMessagePopoverProps> = ({
  username,
  address,
  anchorEl,
  handleClose,
  message,
  createNewContact,
  history,
  users,
  waggleUsers
}) => {
  const open = Boolean(anchorEl)
  const id = open ? 'simple-popover' : undefined
  const registeredUsername = Array.from(Object.values(users)).filter(
    (obj) => obj.address === address
  )[0]
  let waggleIdentity = false
  if (waggleUsers) {
    const arr = Array.from(Object.keys(waggleUsers))
    if (arr.includes(message?.publicKey) || arr.includes(registeredUsername?.publicKey)) {
      waggleIdentity = true
    }
  }
  return (
    <Popover
      id={id}
      open={open}
      anchorEl={anchorEl}
      onClose={handleClose}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'center'
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'center'
      }}
    >
      <QuickActionLayout
        main={username}
        buttonName="Send message"
        handleClose={handleClose}
        warning={
          !waggleIdentity ? 'Unregistered users cannot receive messages.' : null
        }
        onClick={() => {
          if (message?.publicKey || registeredUsername?.publicKey) {
            createNewContact({
              contact: {
                address,
                nickname: username,
                publicKey: message?.publicKey || registeredUsername?.publicKey
              },
              history
            })
          }
        }}
      >
        <Jdenticon size="100" value={username} />
      </QuickActionLayout>
    </Popover>
  )
}

export default SendMessagePopover
