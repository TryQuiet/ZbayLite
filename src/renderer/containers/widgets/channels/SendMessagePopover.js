import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { withRouter } from 'react-router-dom'
import * as R from 'ramda'

import SendMessagePopover from '../../../components/widgets/channels/SendMessagePopover'
import identitySelectors from '../../../store/selectors/identity'
import userSelectors from '../../../store/selectors/users'
import directMessages from '../../../store/handlers/contacts'
import directMessagesSelectors from '../../../store/selectors/directMessages'

export const mapStateToProps = state => ({
  identityId: identitySelectors.id(state),
  users: userSelectors.users(state),
  waggleUsers: directMessagesSelectors.users(state)
})

export const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      createNewContact: (contact) => directMessages.epics.createVaultContact(contact)
    },
    dispatch
  )

export default R.compose(
  React.memo,
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(SendMessagePopover)
