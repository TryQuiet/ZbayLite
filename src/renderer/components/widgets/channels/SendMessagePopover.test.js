import React from 'react'
import { shallow } from 'enzyme'

import { SendMessagePopover } from './SendMessagePopover'
import { mockClasses } from '../../../../shared/testing/mocks'

describe('SendMessagePopover', () => {
  it('renders popover', () => {
    const result = shallow(
      <SendMessagePopover
        classes={mockClasses}
        anchorEl
        handleClose={jest.fn()}
        username='TestUser'
        users={{}}
        address={'ztestsapling1juf4322spfp2nhmqaz5wymw8nkkxxyv06x38cel2nj6d7s8fdyd6dlsmc6efv02sf0kty2v7lfz'}
      />
    )
    expect(result).toMatchSnapshot()
  })
})
