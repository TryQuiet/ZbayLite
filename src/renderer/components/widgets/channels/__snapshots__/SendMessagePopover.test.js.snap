// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMessagePopover renders popover 1`] = `
<WithStyles(ForwardRef(Popover))
  anchorEl={true}
  anchorOrigin={
    Object {
      "horizontal": "center",
      "vertical": "bottom",
    }
  }
  id="simple-popover"
  onClose={[MockFunction]}
  open={true}
  transformOrigin={
    Object {
      "horizontal": "center",
      "vertical": "top",
    }
  }
>
  <Component
    buttonName="Send message"
    handleClose={[MockFunction]}
    main="TestUser"
    onClick={[Function]}
    warning="Unregistered users cannot receive messages."
  >
    <Component
      size="100"
      value="TestUser"
    />
  </Component>
</WithStyles(ForwardRef(Popover))>
`;
